### PR TITLE
Fixes bug where binary operations on PennyLane tensors results in an incorrect requires_grad attribute

### LINKED ISSUE #888: Arithmetic operations between pennylane.numpy arrays have inconsistent `requires_grad` output
**Issue**

The PennyLane wrapper if the numpy library, which is used to optimise circuits with the standard autograd interface, has an inconsistent behaviour when using arithmetic functions such as +, -, * (here using the sum to demonstrate). 

Summing two `pennylane.numpy.array` objects, one having gradients enabled and the other not, returns a differentiable array or not, depending on the order.

``` python
import pennylane as qml
from pennylane import numpy as np

a = np.array(1., requires_grad=True)
b = np.array(0., requires_grad=False)

c = a+b
print(c.requires_grad) # True

d = b+a
print(d.requires_grad) # False
```
**Expected behaviour**
The `requires_grad = True` flag should always be overwriting, and the output of the two print statements should be true in both cases. This is important for example in loss function where target outputs (which do not require gradients) and predictions (which depend on the parameters and require gradients) are compared, i.e. via `np.sum(np.abs((target - prediction))**2)`.

**System information**
Reproducible with PennyLane v0.12 stable release, and with the current master, in tape mode or not in tape mode.

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

When applying Python operators that in turn call an ndarray's `__array_ufunc__` method, the `requires_grad` attribute was not properly propagating to the output:

```pycon
>>> x = np.array([1, 2], requires_grad=True)
>>> y = np.array([2, 3], requires_grad=False)
>>> print("Should be true:", (x - y).requires_grad)
Should be true: True
>>> print("Should be true:", (y - x).requires_grad)
Should be true: False
```

It appears that the `__array_ufunc__` method of the left-hand operator is being called, and as such all arrays are not being taken into account.

**Description of the Change:**

* Adds a `qml.numpy.tensor.__array_ufunc__()` method to correct this behaviour, and take all arrays into account when using Python operators.

**Benefits:**

The bug is now fixed:

```pycon
>>> x = np.array([1, 2], requires_grad=True)
>>> y = np.array([2, 3], requires_grad=False)
>>> print("Should be true:", (x - y).requires_grad)
Should be true: True
>>> print("Should be true:", (y - x).requires_grad)
Should be true: True
```

**Possible Drawbacks:** n/a

**Related GitHub Issues:** Fixes #888

### CODE REVIEW

File: pennylane/numpy/tensor.py [hunk 1]
```diff
@@ -113,6 +113,43 @@ def __array_wrap__(self, obj):
...
+        # pylint: disable=no-member,attribute-defined-outside-init
+        outputs = [
+            i.view(onp.ndarray) if hasattr(i, "unwrap") else i for i in kwargs.get("out", ())
+        ]
+
+        if outputs:
+            outputs = tuple(outputs)
+            kwargs["out"] = outputs
+        else:
+            outputs = (None,) * ufunc.nout
```
Comment (mariaschuld) [inline pennylane/numpy/tensor.py:126]: Hmm, maybe a comment for why the outputs need to be recast this way?

File: pennylane/numpy/tensor.py [hunk 2]
```diff
@@ -113,6 +113,43 @@ def __array_wrap__(self, obj):
         out_arr = tensor(obj, requires_grad=self.requires_grad)
         return super().__array_wrap__(out_arr)
 
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        # pylint: disable=no-member,attribute-defined-outside-init
+        outputs = [
+            i.view(onp.ndarray) if hasattr(i, "unwrap") else i for i in kwargs.get("out", ())
```
Comment (mariaschuld) [inline pennylane/numpy/tensor.py:119]: What would the `hasattr` distinguish here?

File: pennylane/numpy/tensor.py (hunk 1 above)
Comment (mariaschuld) [reply pennylane/numpy/tensor.py:126]: You are just extracting the numpy arrays so you can use the native function, right?

Comment (josh146) [reply pennylane/numpy/tensor.py:126]: Yep, this approach comes from the example shown in the NumPy documentation: https://numpy.org/devdocs/user/basics.subclassing.html#array-ufunc-for-ufuncs

File: pennylane/numpy/tensor.py [hunk 3]
```diff
@@ -113,6 +113,43 @@ def __array_wrap__(self, obj):
...
+            for result, output in zip(res, outputs)
+        ]
+
+        # if any of the inputs were trainable, the output is also trainable
+        requires_grad = any(
+            isinstance(x, onp.ndarray) and getattr(x, "requires_grad", True) for x in inputs
+        )
+
+        for i in range(len(res)):  # pylint: disable=consider-using-enumerate
+            res[i] = tensor(res[i], requires_grad=requires_grad)
```
Comment (mariaschuld) [inline pennylane/numpy/tensor.py:149]: I may be blind, but what happens if the tensor is deeper than 1 dimension? Would we want to wrap each and every subelement?

File: pennylane/numpy/tensor.py (hunk 2 above)
Comment (josh146) [reply pennylane/numpy/tensor.py:119]: You're right, I think we can simply do
```python
outputs = [i.view(onp.ndarray) for i in kwargs.get("out", ())]
```
I'll change it and see if the tests pass.

File: pennylane/numpy/tensor.py [hunk 4]
```diff
@@ -113,6 +113,43 @@ def __array_wrap__(self, obj):
...
+
+        return res[0] if len(res) == 1 else tuple(res)
```
Comment (mariaschuld) [inline pennylane/numpy/tensor.py:151]: Arg I am really not clued up on this module...why would you return a tuple here? Not the onp.array?

File: pennylane/numpy/tensor.py (hunk 3 above)
Comment (josh146) [reply pennylane/numpy/tensor.py:149]: Oh, `res` is a flat list of tensors! To cover for NumPy ufuncs that return more than one output. So each element of the list corresponds to a potentially multi-dimensional tensor.

File: pennylane/numpy/tensor.py (hunk 4 above)
Comment (josh146) [reply pennylane/numpy/tensor.py:151]: This is a restriction on NumPy ufuncs - they either return a single numpy array, or a tuple of arrays. This echoes how Python functions work.

Comment (josh146) [reply pennylane/numpy/tensor.py:151]: I'll add comments

Comment (josh146) [reply pennylane/numpy/tensor.py]: Have added some comments, let me know if it makes it easier to follow!

File: pennylane/numpy/tensor.py (hunk 2 above)
Comment (josh146) [reply pennylane/numpy/tensor.py]: it works!

File: pennylane/numpy/tensor.py (hunk 1 above)
Comment (josh146) [reply pennylane/numpy/tensor.py:130]: Done!

File: tests/test_numpy_wrapper.py [hunk 5]
```diff
@@ -350,6 +350,58 @@ def test_binary_operator_nontrainable(self):
         assert isinstance(res, np.tensor)
         assert not res.requires_grad
 
+    def test_binary_operator_mixed_trainable_left(self):
+        """Test that binary operators on one trainable and one non-trainable
+        arrays result in trainable output."""
+        x = np.array([[1, 2], [3, 4]], requires_grad=True)
+        y = np.array([[5, 6], [7, 8]], requires_grad=False)
```
Comment (mariaschuld) [inline tests/test_numpy_wrapper.py:357]: So the elements inside a trainable array will become trainable when indexed like `x[0]`, right? Is this why we only need to check that the overall output array is trainable?

I'm thinking of an edge case `x =  np.array([np.array([1, 2], requires_grad=False), [3, 4]], requires_grad=True)` here

Comment (josh146) [reply tests/test_numpy_wrapper.py:357]: Should do!

image

**Review by mariaschuld — APPROVED** [review]
I am not clued up on this method that is added here, but for all I can see it looks good. 

@josh146 up to your judgement if you need a second reviewer. But since this is an important bugfix I am not opposed to merging.

File: pennylane/numpy/tensor.py (hunk 4 above)
Comment (mariaschuld) [reply pennylane/numpy/tensor.py]: Yes, thanks!
